src/melee.cpp

```diff
@@ -808,10 +818,16 @@ bool Character::melee_attack_abstract( Creature &t, bool allow_special,
             weakpoint_attack attack;
             attack.weapon = &cur_weap;
             t.deal_melee_hit( this, hit_spread, critical_hit, d, dealt_dam, attack, &target_bp );
-            if( dealt_special_dam.type_damage( damage_type::CUT ) > 0 ||
-                dealt_special_dam.type_damage( damage_type::STAB ) > 0 ||
-                ( !cur_weapon && ( dealt_dam.type_damage( damage_type::CUT ) > 0 ||
-                                   dealt_dam.type_damage( damage_type::STAB ) > 0 ) ) ) {
+
+            bool has_edged_damage = false;
+            for( const damage_type &dt : damage_type::get_all() ) {
+                if( dt.melee_only && dt.edged && dealt_special_dam.type_damage( dt.id ) > 0 ) {
+                    has_edged_damage = true;
+                    break;
+                }
+            }
+
+            if( has_edged_damage || ( !cur_weapon && has_edged_damage ) ) {
```

Comment:
This looks like there is a logical error in here - you check for `has_edged_damage` twice, so that check could be simply  omitted.